index.bs

```diff
@@ -249,69 +530,98 @@ A <dfn export>file locator</dfn> is a [=/file system locator=] whose
+    to |childPath|'s [=list/size=], exclusive,
+    [=list/append=] |childPath|[|index|] to |relativePath|.
+
+1. Return |relativePath|.
+
+Note: These steps have to be run on the [=file system queue=].
 
 </div>
 
 <div algorithm>
-The <dfn export data-lt="getting the locator">get the locator</dfn> algorithm given
-[=/file system entry=] |entry| runs an [=implementation-defined=] series of steps adhering to these
-constraints:
+To <dfn for="file system locator">get the relationship</dfn> between a
+[=/file system locator=] |self| to a [=file system locator=] |related|:
+
+1. If |self|'s [=file system locator/file system=]  is not |related|'s
+   [=file system locator/file system=], return "`other`".
+
+1. Let |selfPath| be |self|'s [=file system locator/path=].
+1. Let |relatedPath| be |related|'s [=file system locator/path=].
+
+1. Let |selfPathSize| be |selfPath|'s [=list/size=].
+1. Let |relatedPathSize| be |relatedPath|'s [=list/size=].
 
-- If |entry| is a [=file entry=], they return a [=file locator=].
-- If |entry| is a [=directory entry=], they return a [=directory locator=].
-- If these steps return |locator|, then:
-  - [=Locating an entry=] with |locator| returns |entry|,
-    provided no intermediate file system operations were run.
-  - |entry|'s [=file system entry/name=] is the last [=list/item=] of |locator|'s
-    [=file system locator/path=].
+1. [=list/For each=] |index| of |selfPath|'s [=list/indices=]:
+  1. If |index| is greater than or equal to |relatedPathSize|, return "`ancestor`".
+  1. If |selfPath|[|index|] is not |relatedPath|[|index|], return "`other`".
+
+1. If |selfPathSize| equals |relatedPathSize|, return "`self`".
+1. If |selfPathSize| + 1 equals |relatedPathSize|, return "`direct child`".
+1. Return "`descendant`".
+
+Note: These steps have to be run on the [=file system queue=].
 
 </div>
 
-A <dfn export>file system path</dfn> is a [=/list=] of one or more [=strings=].
-This may be a virtual path that is mapped to real location on disk or in memory,
-may correspond directly to a path on the local file system, or may not
-correspond to any file on disk at all. The actual physical location of the
-corresponding [=/file system entry=] is [=implementation-defined=].
+<div algorithm>
+The <dfn export data-lt="locating an entry">locate an entry</dfn> algorithm given a
+[=/file system locator=] |locator| runs the following steps:
 
-<p class=example id=example-path>Let |path| be the [=/list=]
-« "`data`", "`drafts`", "`example.txt`" ».
-There is no expectation that a file named `example.txt` exists anywhere on disk.
+1. Let |fileSystem| be |locator|'s [=file system locator/file system=].
+1. Let |path| be |locator|'s [=file system locator/path=].
+1. Let |entry| be the result of running |fileSystem|'s
+   [=file system/locate an entry=] given |path|.
+1. If |entry| is null, return null.
+1. If |locator| is a [=file locator=], [=Assert=]: |entry| is a [=file entry=].
```

Comment:
Are these "Assert"s actually valid as assertions? As far as I can tell it is totally possible for this to be called with a locator whose type doesn't match the entry that exists now (for example if you have a FileHandle with a certain locator, delete the entry, re-create it as a directory, and then try to do anything with the file handle).